tests/components/neato/test_config_flow.py

```diff
@@ -1,160 +1,166 @@
+from homeassistant.components.neato.config_flow import OAuth2FlowHandler
+from homeassistant.components.neato.const import NEATO_DOMAIN
+from homeassistant.helpers import config_entry_oauth2_flow
 
 from tests.async_mock import patch
 from tests.common import MockConfigEntry
+from tests.helpers.test_config_entry_oauth2_flow import MockOAuth2Implementation
+from tests.helpers.test_config_entry_oauth2_flow import local_impl  # noqa: F401
+
+CLIENT_ID = "1234"
+CLIENT_SECRET = "5678"
+
+VENDOR = Neato()
+OAUTH2_AUTHORIZE = VENDOR.auth_endpoint
+OAUTH2_TOKEN = VENDOR.token_endpoint
+
+
+async def test_full_flow(
+    hass, aiohttp_client, aioclient_mock, current_request_with_host
+):
+    """Check full flow."""
+    assert await setup.async_setup_component(
+        hass,
+        "neato",
+        {
+            "neato": {"client_id": CLIENT_ID, "client_secret": CLIENT_SECRET},
+            "http": {"base_url": "https://example.com"},
+        },
+    )
 
-USERNAME = "myUsername"
-PASSWORD = "myPassword"
-VENDOR_NEATO = "neato"
-VENDOR_VORWERK = "vorwerk"
-VENDOR_INVALID = "invalid"
+    result = await hass.config_entries.flow.async_init(
+        "neato", context={"source": config_entries.SOURCE_USER}
+    )
+    state = config_entry_oauth2_flow._encode_jwt(
+        hass,
+        {
+            "flow_id": result["flow_id"],
+            "redirect_uri": "https://example.com/auth/external/callback",
+        },
+    )
 
+    assert result["url"] == (
+        f"{OAUTH2_AUTHORIZE}?response_type=code&client_id={CLIENT_ID}"
+        "&redirect_uri=https://example.com/auth/external/callback"
+        f"&state={state}"
+        f"&client_secret={CLIENT_SECRET}"
+        "&scope=public_profile+control_robots+maps"
+    )
+
+    client = await aiohttp_client(hass.http.app)
+    resp = await client.get(f"/auth/external/callback?code=abcd&state={state}")
+    assert resp.status == 200
+    assert resp.headers["content-type"] == "text/html; charset=utf-8"
+
+    aioclient_mock.post(
+        OAUTH2_TOKEN,
+        json={
+            "refresh_token": "mock-refresh-token",
+            "access_token": "mock-access-token",
+            "type": "Bearer",
+            "expires_in": 60,
+        },
+    )
+
+    with patch(
+        "homeassistant.components.neato.async_setup_entry", return_value=True
+    ) as mock_setup:
+        await hass.config_entries.flow.async_configure(result["flow_id"])
 
-@pytest.fixture(name="account")
-def mock_controller_login():
-    """Mock a successful login."""
-    with patch("homeassistant.components.neato.config_flow.Account", return_value=True):
-        yield
+    assert len(hass.config_entries.async_entries(NEATO_DOMAIN)) == 1
+    assert len(mock_setup.mock_calls) == 1
 
 
-def init_config_flow(hass):
-    """Init a configuration flow."""
-    flow = config_flow.NeatoConfigFlow()
+async def test_abort_if_already_setup(hass):
+    """Test we abort if Neato is already setup."""
+    flow = OAuth2FlowHandler()
```

Comment:
Please start the flow with `hass.config_entries.flow.async_init` and continue the flow with `hass.config_entries.flow.async_configure`.

Reply:
Done. The last test still calls the tested function directly, because I want to test this function independently.